File: best-time-to-buy-and-sell-stock.cpp
//Time - O(n)
//Space - O(1)
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int low = prices[0];
        int profit = INT_MIN;
        for(auto p:prices){
            if(low>p){
                low = p;
            }else{
                profit = max(profit, p-low);
            }
        }
        
        return profit;
    }
};

File: best_time_to_buy_and_sell_stock.java
// Time Complexity : O(n)
// Space Complexity : O(1)

class Solution {
    public int maxProfit(int[] prices) {
        
        int profit = Integer.MIN_VALUE;
        int buy = Integer.MAX_VALUE;
        
        // Iterate through each element
        for (int i =0; i < prices.length ; i++){
            
            // Store the buy with the lowest price 
            buy = Math.min(buy, prices[i]);
            
            // Store the max profit based on the latest buy
            profit = Math.max(profit , prices[i] - buy);
            
        }
        
        return profit;
        
    }
}